### PR TITLE
PROPOSAL: add preload and library type

### PR DESCRIPTION
**Pull request by MementoRC**
Can we add recording of the version and the type of library coincurve was built upon (external or internal)? In the case of external, this information is needed to try to pre-load the DLL on windows. This is thw main reason why the tests are not working in the othe PR

### CODE REVIEW

**Comment by MementoRC** [comment]
@ofek Do you accept this proposal, or should I try to find another way. This is what I need to force the loading of `libsecp256k1-2.dll` for the windows wheel build with the conda install. So far what I observed is that:
   - No particular issues with building, we need conda to install libsecp256k1, no issues
   - For the test of the wheel, cibuildwheel uses virtualenv to install the wheel, but we need conda and things go south from there, the lib is installed, but virtualenv does not see it, only a manual load make the test work
   - 

I think it is valuable to have it in coincurve. The type of build  and the libname (-1, -2, ..) is saved in this file and packaged with the library. When a user initialize the package, it does this simple check as to how the _libsecp256k1 was built and requires the dynamic lib if needed

**Comment by ofek** [comment]
Can you show me another example of an extension module that has runtime logic specifically for Conda?

**Comment by MementoRC** [comment]
No, I don't think I can. This is not specific to conda though, I just know that I will need it for conda, which is the only known case for using libsecp256k1 shared library so far

**Comment by ofek** [comment]
I'm not inherently against it, just trying to understand because I have limited time to review

File: coincurve/__init__.py [hunk 1]
```diff
@@ -1,9 +1,44 @@
-from coincurve.context import GLOBAL_CONTEXT, Context
-from coincurve.keys import PrivateKey, PublicKey, PublicKeyXOnly
-from coincurve.utils import verify_signature
+import ctypes
+import os
+import platform
+import warnings
+
+from coincurve._secp256k1_library_info import SECP256K1_LIBRARY_NAME, SECP256K1_LIBRARY_TYPE
```
Comment (ofek) [inline coincurve/__init__.py:6]: Please use lazy imports by moving all of these into the function and after the possible return statements.

**Review by ofek — CHANGES_REQUESTED** [review]

**Comment by MementoRC** [comment]
I thought about it over lunch and it seems that it is only needed in a niche case where (as far as I understand it) an installed library on windows is "lost". This seems to happen only in the additional condition that several unix-like environments are stacked

So, although I am pretty happy with having found and resolved this flaw, it may just seeds an  unforeseen issue later that may reduce the popularity of `coincurve` ... and I wouldn't want to again be the source of an issue

**Comment by ofek** [comment]
So do you mean that we should close this for now since it's mostly for debugging an edge case that you have fixed?

**Comment by MementoRC** [comment]
Well, if you feel that it is risky, or even feel uneasy about it. I'd rather go back to the drawing board and find another solution or simply test in my fork then PR a clean code without this test

I also don't mean to use too much of your time, but I also need to know what is a good idea to pursuit and what is not

**Comment by ofek** [comment]
Sorry I was just a bit confused: is this necessary to get something working or is it more for debugging? If the former then I will merge.

**Comment by MementoRC** [comment]
It is necessary to make the `cibuildwheel` + `coincurve built with shared libsecp256k1 library` be able to test that wheel on windows. It is a narrow and niche case need

That need "could" occur IF a user happens to install `coincurve` in an unconventional way, with `libsecp256k1` from conda and source-build package with `pip` ... maybe.

**Comment by MementoRC** [comment]
@ofek Are you open to reorganizing `coincurve` as a src-layout package ... I get a lot of missing coincurve package when a `tox -e ...` fails

File: coincurve/__init__.py [hunk 2]
```diff
@@ -1,6 +1,41 @@
-from coincurve.context import GLOBAL_CONTEXT, Context
-from coincurve.keys import PrivateKey, PublicKey, PublicKeyXOnly
-from coincurve.utils import verify_signature
+def load_secp256k1_conda_library():
+    """Load the secp256k1 library from the conda environment."""
+    import warnings
```
Comment (ofek) [inline coincurve/__init__.py:3]: Let's move this inside the conditions

File: coincurve/__init__.py [hunk 3]
```diff
@@ -1,6 +1,41 @@
...
+
+    from coincurve._secp256k1_library_info import SECP256K1_LIBRARY_NAME, SECP256K1_LIBRARY_TYPE
+
+    if SECP256K1_LIBRARY_TYPE != 'EXTERNAL':
```
Comment (ofek) [inline coincurve/__init__.py:7]: Isn't internal the expected default and so this would happen for most users?

**Review by ofek — CHANGES_REQUESTED** [review]

File: coincurve/__init__.py (hunk 3 above)
Comment (MementoRC) [reply coincurve/__init__.py:7]: Right, it should only do extra when coincurve was built with a system library. I need to clean up my debugging.
I am off for the next 10hrs

File: coincurve/__init__.py (hunk 2 above)
Comment (MementoRC) [reply coincurve/__init__.py:3]: I don't understand what you mean

File: coincurve/__init__.py (hunk 3 above)
Comment (MementoRC) [reply coincurve/__init__.py:7]: Let me take another look and use `ctypes.util.find_library()` and only add conda when the library is not found, then fail if it is not a conda environment

Comment (MementoRC) [reply coincurve/__init__.py:7]: I reorganized it a little (it works correctly well on my testing branch). I added the build with conda for verification. I am working on an attempt to test with a macos build, if that works, I guess the linux build should also be manageable. The idea would be to install conda on the docker VM used for  the build in cibuildwheel

File: coincurve/__init__.py (hunk 2 above)
Comment (MementoRC) [reply coincurve/__init__.py:3]: Oh right got it, I was not clear of debug  ... I am a messy eater :(

File: coincurve/__init__.py [hunk 4]
```diff
@@ -1,6 +1,43 @@
-from coincurve.context import GLOBAL_CONTEXT, Context
-from coincurve.keys import PrivateKey, PublicKey, PublicKeyXOnly
-from coincurve.utils import verify_signature
+def load_secp256k1_conda_library():
+    """Load the secp256k1 library from the conda environment."""
+    import warnings
+
```
Comment (ofek) [inline coincurve/__init__.py:4]: Let's move this to when the exception occurs

**Review by ofek — CHANGES_REQUESTED** [review]

**Comment by MementoRC** [comment]
Well, of course - I am switching to dbg mode on my fork - back when it passes

File: coincurve/__init__.py (hunk 4 above)
Comment (MementoRC) [reply coincurve/__init__.py:4]: Yes, I just saw it, working on 2 versions with my nut-sized brain never works well - The gist of it s there so you can continue reviewing, I need to debug why the build does not find the lib - I swear it works fine on my messy branch

Comment (ofek) [reply coincurve/__init__.py]: I don't care at all about failing intermediate commits as long as the final one is green 😄 

I do, as you've seen, care about keeping the import time to an absolute minimum so this library plays nicely with CLIs.

Comment (MementoRC) [reply coincurve/__init__.py]: Rightly so. Initially, pycharm tooltip recommended to have the imports at the top, so I adopted this as a default, in fact a lot of time I use the tool to find and select the import and it places them at the top
I am glad that I now have a good example of when it is better to fine-tune the imports

It should also be better to just import what is needed, or is it less important, like `from lib import the_fct_I_need`, at least I feel it should have a better memory profile

**Comment by MementoRC** [comment]
Ok, I stupidly am going through everysingle fail I solved before ... :'( - This one is the pkg-config redefining the path. Next commit should be the charm

**Comment by ofek** [comment]
Let me know when to review what

**Comment by MementoRC** [comment]
Sure, I still need to work on rebasing this one. Note that I changed the logic, now I only add the extra file to the wheel if it is an EXTERNAL libsecp256k1, then I try/except on its existence to quick-turn if there's no file

**Comment by MementoRC** [comment]
@ofek Do you understand the codecov error?

**Comment by MementoRC** [comment]
@ofek Do you understand the codecov error? I added a new test for the __init__.py logic

**Comment by MementoRC** [comment]
Ok, so I gave some thoughts to this PR. It is solving a niche case and working around the shortcomings of cibuildwheel. In the meantime, when working on testing the coicurve conda environment, I think that I solved the problem in a different way.

So the only benefit of this PR is for the case where coincurve is built to use a system/conda library, but when it is used that library cannot be found - This is a case that is hard to imagine ever happening in the field, and even so, it would only change one error message for another

In brief, I now think it is at best useless.
